mevcommitclient/mev_commit_client.go

```diff
@@ -0,0 +1,136 @@
+	builderRegistryCaller      *builderRegistry.ProviderregistryCaller
+	builderRegistryFilterer    *builderRegistry.ProviderregistryFilterer
+	l1Client                   *ethclient.Client
+	mevCommitClient            *ethclient.Client
+	contractAbi                abi.ABI
+}
+
+const (
+	abiJSON = `[{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"builder","type":"address"},{"indexed":false,"internalType":"uint256","name":"value","type":"uint256"},{"indexed":false,"internalType":"bytes","name":"blsPublicKey","type":"bytes"}],"name":"BuilderRegistered","type":"event"},{"inputs":[{"internalType":"address","name":"builder","type":"address"}],"name":"isBuilderValid","outputs":[],"stateMutability":"view","type":"function"}]`
+)
+
+func NewMevCommitClient(l1MainnetUrl, mevCommitUrl string, validatorRouterAddress, builderRegistryAddress common.Address) (IMevCommitClient, error) {
+	l1Client, err := ethclient.Dial(l1MainnetUrl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to L1 Mainnet: %w", err)
+	}
+
+	mevCommitClient, err := ethclient.Dial(mevCommitUrl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to MEV-Commit EVM: %w", err)
+	}
+
+	validatorOptInRouter, err := validatoroptinrouter.NewValidatoroptinrouterCaller(validatorRouterAddress, l1Client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create ValidatorOptInRouter caller: %w", err)
+	}
+
+	builderRegistryCaller, err := builderRegistry.NewProviderregistryCaller(builderRegistryAddress, mevCommitClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create BuilderRegistry caller: %w", err)
+	}
+
+	builderRegistryFilterer, err := builderRegistry.NewProviderregistryFilterer(builderRegistryAddress, mevCommitClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create BuilderRegistry filterer: %w", err)
+	}
+
+	contractAbi, err := abi.JSON(strings.NewReader(abiJSON))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse contract ABI: %w", err)
+	}
+
+	return &MevCommitClient{
+		L1Address:                  l1MainnetUrl,
+		MevCommitAddress:           mevCommitUrl,
+		ValidatorRouterAddress:     validatorRouterAddress,
+		BuilderRegistryAddress:     builderRegistryAddress,
+		validatorOptInRouterCaller: validatorOptInRouter,
+		builderRegistryCaller:      builderRegistryCaller,
+		builderRegistryFilterer:    builderRegistryFilterer,
+		l1Client:                   l1Client,
+		mevCommitClient:            mevCommitClient,
+		contractAbi:                contractAbi,
+	}, nil
+}
+
+func (m *MevCommitClient) GetOptInStatusForValidators(pubkeys [][]byte) ([]bool, error) {
+	return m.validatorOptInRouterCaller.AreValidatorsOptedIn(nil, pubkeys)
+}
+func (m *MevCommitClient) GetActiveBuilders() ([][]byte, error) {
+	opts := &bind.FilterOpts{
+		Start: 0,
+		End:   nil, // Latest block
+	}
+
+	iterator, err := m.builderRegistryFilterer.FilterProviderRegistered(opts, nil)
```

Comment:
Have you ran this against the testnet yet? Most RPCs (I think for our chain too) have a limit of 50k blocks for querying events, hence this would require querying in 50k batches.

Reply:
I've run it agains't testnet, but we should prob opt for a chaching strategy of some sort.